db_path_ezvir=config["ezvir_database"]
rule all_ezvir:
    input: expand("Ezvir/test_cut-0_p1-plot.pdf",sample=list(read_naming.keys()))


rule map_paired_reads:
    conda: pipeline_path + "envs/ezvir.yml"

    input: r1=expand("unmapped/{host_acc}/{sample}_R1.fq",host_acc=config['reference_genome']['filename'],sample=list(read_naming.keys())),
            r2=expand("unmapped/{host_acc}/{sample}_R2.fq",host_acc=config['reference_genome']['filename'],sample=list(read_naming.keys()))

    params: genome_path=db_path_ezvir + "index/{genome_id}"

    threads: 10

    output: "Ezvir/alignments/{genome_id}.bam"

    shell:
        """
        bowtie2 -p {threads} -q1 {input.r1} -q2 {input.r2} \
        -x {params.genome_path} | samtools view -bS -F 12 | samtools sort -n -O BAM -o {output} 
        """

rule calculate_genome_coverage:
    conda: pipeline_path + "envs/ezvir.yml"

    input: "Ezvir/alignments/{genome_id}.bam"

    output: "Ezvir/coverage/{genome_id}.csv"

    shell: "bedtools genomecov -d -ibam {input} > {output}"


rule phase_1_report:
    conda: pipeline_path + "envs/ezvir.yml"

    input: "Ezvir/coverage/{genome_id}.csv"

    output: "Ezvir/test_cut-0_p1-plot.pdf",
            "Ezvir/test_cut-0_p1-values.txt",
            "Ezvir/BEST-GENOMES.ezv",
            "Ezvir/ALL-RESULTS.ezv"


    params: script_path=pipeline_path + "rules/taxonomy_profiling/scripts/Ezvir"

    shell: "{params.script_path}/Report.r -c Ezvir/COLORS.ezv -j GENOMES_NO-ANIMAL-140527.ezv -i {input} -o test > {output}"




