

rule combine_contig_annotations:
    conda:
        "../../envs/python-r.yml"
    singularity:
        "docker://metagenlab/microbiome-shotgun-pipeline:1.0"
    input:
        rgi = expand("samples/{sample}/annotation/rgi/rgi_proteins.json", sample = list(read_naming.keys())),
        core_genes = expand("samples/{sample}/contigs_classification/core_genes/{sample}.tsv", sample = list(read_naming.keys())),
        plsdb = expand("samples/{sample}/contigs_classification/plsdb/{sample}.tsv", sample = list(read_naming.keys())),
        COG_mobilome = expand("samples/{sample}/contigs_classification/COG_mobilome/{sample}.tsv", sample = list(read_naming.keys())),
        deepvirfinder = expand("samples/{sample}/contigs_classification/deepvirfinder/large_contigs_edit.fasta_gt500bp_dvfpred.txt", sample = list(read_naming.keys())),
        depth = expand("samples/{sample}/bwa/{sample}_assembled/{sample}_metabat2_depth.txt", sample = list(read_naming.keys())),
        cat_taxonomy = expand("samples/{sample}/contigs_classification/cat_bat/out.CAT.ORF2LCA.txt", sample = list(read_naming.keys())),
        GC = expand("samples/{sample}/contigs_classification/{sample}_GC.tsv", sample = list(read_naming.keys())),
    params:
        plasmid_table = "databases/plasmids/plsdb.tsv"
    output:
       "report/contigs_classification/merged.db" 
    script: "scripts/combine.py"