

rule extract_read_counts:
    input:
        "samples/{sample}/reads/trimmed/fastqc/{pair}_paired_fastqc/fastqc_data.txt",
    output:
        "samples/{sample}/reads/trimmed/fastqc/{pair}_paired_fastqc/{pair}_reads_count.txt"
    shell:
       """
       echo sed 's/.*:Total Sequences\\t//'
       grep "Total Sequences" {input[0]} | sed 's/.*:Total Sequences\\t//' > {output[0]}
       """

rule calculate_RPKM:
    conda:
        "../../envs/python-r.yml"
    singularity:
        "docker://metagenlab/microbiome-shotgun-pipeline:1.0"
    params:
        config["local_samples"],
    input:
        sample_list = expand("samples/{sample}/{{data_type}}/{{db_name}}/{{search_tool}}_best_hits_{{identity}}.m8", sample=read_naming.keys()),
        flash_log = expand("samples/{sample}/reads/raw/{sample}.log", sample=read_naming.keys()),
        reference_fasta = "reference_databases/{data_type}/{db_name}.faa",
    output:
        "report/{data_type}/{db_name}/{search_tool}_RPKM_{identity,[0-9]+}.tsv"
    script: "scripts/calculate_RPKM.py"


rule RPKM_table_add_annotation_virulence:
    conda:
        "../../envs/python-r.yml"
    singularity:
        "docker://metagenlab/microbiome-shotgun-pipeline:1.0"
    input:
        virulence_database = "databases/virulence/sqlite_db/virulence.db",
        RPKM_database = "report/{search_tool}/virulence/VF_db/RPKM.db",
    output:
        "report/{search_tool}/virulence/VF_db/annotation.log"
    script: "scripts/virulence_annotation.py"


rule RPKM_table_add_annotation_rgi:
    conda:
        "../../envs/python-r.yml"
    singularity:
        "docker://metagenlab/microbiome-shotgun-pipeline:1.0"
    input:
        RPKM_table = "report/resistance/CARD_protein_homolog_model_edit/{search_tool}_RPKM_{identity}.tsv",
        card_annotation = "reference_databases/resistance/accession2AMR_family.tab",
        silix_98 = "reference_databases/resistance/CARD_protein_homolog_model_edit_0.98_0.6.fnodes",
        silix_95 = "reference_databases/resistance/CARD_protein_homolog_model_edit_0.95_0.6.fnodes",
        silix_90 = "reference_databases/resistance/CARD_protein_homolog_model_edit_0.90_0.6.fnodes",
        silix_80 = "reference_databases/resistance/CARD_protein_homolog_model_edit_0.80_0.6.fnodes",
        silix_98_annotation = "reference_databases/resistance/CARD_protein_homolog_model_edit_0.98_0.6_annotation.tsv",
        silix_95_annotation = "reference_databases/resistance/CARD_protein_homolog_model_edit_0.95_0.6_annotation.tsv",
        silix_90_annotation = "reference_databases/resistance/CARD_protein_homolog_model_edit_0.90_0.6_annotation.tsv",
        silix_80_annotation = "reference_databases/resistance/CARD_protein_homolog_model_edit_0.80_0.6_annotation.tsv",
    output:
        # report/resistance/CARD_protein_homolog_model_edit/diamond_RPKM_90_annotated.tsv
        "report/resistance/CARD_protein_homolog_model_edit/{search_tool}_RPKM_{identity,[0-9]+}_annotated.tsv"
    script: "scripts/resistance_annotation.py"
