rule prepare_contig_bins_table:
    conda:
        "../../envs/python-r.yml"
    singularity:
        "docker://metagenlab/microbiome-shotgun-pipeline:1.0"
    input:
        fasta_files = dynamic("samples/{sample}/binning/maxbin/{bin_id}.fasta"),
    output:
        "samples/{sample}/binning/maxbin/maxbin_table.tab",
    log:
        "samples/{sample}/binning/maxbin/{sample}.log",
    script:
        'scripts/maxbin2classif.py'


rule plot_gc_cov_maxbin_bins:
    conda:
        "../../envs/python-r.yml"
    singularity:
        "docker://metagenlab/microbiome-shotgun-pipeline:1.0"
    input:
        "samples/{sample}/binning/maxbin/maxbin_table.tab",
        "samples/{sample}/binning/maxbin/contig_abundance.txt",
        "samples/{sample}/assembly/spades/large_contigs_edit.fasta",
    output:
        "samples/{sample}/binning/maxbin/gc_cov_buble_test.svg"
    script:
        'scripts/gc_coverage_plot.py'


'''
rule plot_gc_cov_kaiju_phylum:
    input:

    output:

    shell:

        scripts/gc_coverage_plot.py -t maxbin_table.tab -m contigs_500bp_renamed.fasta -d contig_abundance.txt


# extract contig name from prodigal output
rule kaiju_phylum_contig_consensus:
    input:

    output:

    shell:

        scripts/get_contig2majority_kaiju_phylum.py -g RHT_04182018.gbf -l locus_taxonomy.tab > contig_taxonomy.tab

'''
